cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go

```diff
@@ -112,63 +126,116 @@ func createKubeConfigFiles(outDir string, cfg *kubeadmapi.InitConfiguration, kub
+		return err
 	}
 
+	for _, kubeConfigFileName := range kubeConfigFileNames {
+		// retrieves the KubeConfigSpec for given kubeConfigFileName
+		spec, exists := specs[kubeConfigFileName]
+		if !exists {
+			return errors.Errorf("couldn't retrieve KubeConfigSpec for %s", kubeConfigFileName)
+		}
+
+		// builds the KubeConfig object
+		config, csr, err := buildKubeConfigAndCSRFromSpec(spec, cfg.ClusterName)
+		if err != nil {
+			return err
+		}
+
+		// writes the kubeconfig to disk if it not exists
+		if err = createKubeConfigFileIfNotExists(kubeConfigDir, kubeConfigFileName, config); err != nil {
+			return err
+		}
+		// Write CSR to disk
+		if err := pkiutil.WriteCSR(kubeConfigDir, kubeConfigFileName, csr); err != nil {
+			return errors.Wrap(err, "failed to write CSR file")
+		}
+	}
+
+	return nil
+}
+
+// getKubeConfigSpecs returns all KubeConfigSpecs actualized to the context of the current InitConfiguration
+// NB. this methods holds the information about how kubeadm creates kubeconfig files.
+func getKubeConfigSpecsBase(cfg *kubeadmapi.InitConfiguration) (map[string]*kubeConfigSpec, error) {
 	controlPlaneEndpoint, err := kubeadmutil.GetControlPlaneEndpoint(cfg.ControlPlaneEndpoint, &cfg.LocalAPIEndpoint)
 	if err != nil {
 		return nil, err
 	}
 
 	var kubeConfigSpec = map[string]*kubeConfigSpec{
 		kubeadmconstants.AdminKubeConfigFileName: {
-			CACert:     caCert,
 			APIServer:  controlPlaneEndpoint,
 			ClientName: "kubernetes-admin",
 			ClientCertAuth: &clientCertAuth{
-				CAKey:         caKey,
 				Organizations: []string{kubeadmconstants.SystemPrivilegedGroup},
 			},
 		},
 		kubeadmconstants.KubeletKubeConfigFileName: {
-			CACert:     caCert,
 			APIServer:  controlPlaneEndpoint,
 			ClientName: fmt.Sprintf("%s%s", kubeadmconstants.NodesUserPrefix, cfg.NodeRegistration.Name),
 			ClientCertAuth: &clientCertAuth{
-				CAKey:         caKey,
 				Organizations: []string{kubeadmconstants.NodesGroup},
 			},
 		},
 		kubeadmconstants.ControllerManagerKubeConfigFileName: {
-			CACert:     caCert,
-			APIServer:  controlPlaneEndpoint,
-			ClientName: kubeadmconstants.ControllerManagerUser,
-			ClientCertAuth: &clientCertAuth{
-				CAKey: caKey,
-			},
+			APIServer:      controlPlaneEndpoint,
+			ClientName:     kubeadmconstants.ControllerManagerUser,
+			ClientCertAuth: &clientCertAuth{},
 		},
 		kubeadmconstants.SchedulerKubeConfigFileName: {
-			CACert:     caCert,
-			APIServer:  controlPlaneEndpoint,
-			ClientName: kubeadmconstants.SchedulerUser,
-			ClientCertAuth: &clientCertAuth{
-				CAKey: caKey,
-			},
+			APIServer:      controlPlaneEndpoint,
+			ClientName:     kubeadmconstants.SchedulerUser,
+			ClientCertAuth: &clientCertAuth{},
 		},
 	}
 
 	return kubeConfigSpec, nil
 }
 
+// getKubeConfigSpecs returns all KubeConfigSpecs actualized to the context of the current InitConfiguration
+// NB. this methods holds the information about how kubeadm creates kubeconfig files.
```

Comment:
we seem to have the same NB line here and for getKubeConfigSpecsBase.
should we only have it for one of the functions?